File: pipelinereposerver/src/main/java/com/greenops/pipelinereposerver/api/PipelineRepoApi.java
package com.greenops.pipelinereposerver.api;

import com.greenops.pipelinereposerver.repomanager.RepoManager;
import com.greenops.util.datamodel.git.GitRepoSchema;
import com.greenops.util.datamodel.git.GitRepoSchemaInfo;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@Slf4j
@RestController
@RequestMapping("/repo")
public class PipelineRepoApi {

    private final RepoManager repoManager;

    @Autowired
    public PipelineRepoApi(RepoManager repoManager) {
        this.repoManager = repoManager;
    }

    @PostMapping(value = "clone/{org}")
    ResponseEntity<Void> cloneRepo(@PathVariable("org") String org, @RequestBody GitRepoSchema gitRepoSchema) {
        if (!repoManager.getOrgName().equals(org)) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
        }
        if (repoManager.containsGitRepoSchema(gitRepoSchema)) {
            if (repoManager.update(gitRepoSchema)) {
                return ResponseEntity.ok().build();
            } else {
                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
            }
        } else {
            if (repoManager.clone(gitRepoSchema)) {
                return ResponseEntity.ok().build();
            } else {
                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
            }
        }
    }

    @PostMapping(value = "delete")
    ResponseEntity<Void> deleteRepo(@RequestBody GitRepoSchema gitRepoSchema) {
        if (repoManager.delete(gitRepoSchema)) {
            return ResponseEntity.ok().build();
        } else {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    @PostMapping(value = "sync")
    ResponseEntity<String> syncRepo(@RequestBody GitRepoSchema gitRepoSchema) {
        var revisionHash = repoManager.sync(gitRepoSchema);
        if (revisionHash != null) {
            return ResponseEntity.ok(revisionHash);
        } else {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    @PostMapping(value = "resetToVersion/{orgName}/{teamName}/{gitCommit}")
    ResponseEntity<Void> resetRepoToVersion(@PathVariable("orgName") String orgName,
                                            @PathVariable("teamName") String teamName,
                                            @PathVariable("gitCommit") String gitCommit,
                                            @RequestBody GitRepoSchemaInfo gitRepoSchemaInfo) {
        var gitRepoSchema = new GitRepoSchema(gitRepoSchemaInfo.getGitRepo(), gitRepoSchemaInfo.getPathToRoot(), null);
        if (repoManager.getOrgName().equals(orgName) && repoManager.containsGitRepoSchema(gitRepoSchema)) {
            if (repoManager.resetToVersion(gitCommit, gitRepoSchema)) {
                return ResponseEntity.ok().build();
            } else {
                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
            }
        } else {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
        }
    }
}

File: pipelinereposerver/api/pipeline_repo_api.go
package api

import (
	"bytes"
	"github.com/gorilla/mux"
	"github.com/greenopsinc/pipelinereposerver/repomanager"
	"github.com/greenopsinc/util/db"
	"github.com/greenopsinc/util/git"
	"github.com/greenopsinc/util/kubernetesclient"
	"net/http"
	"os"
)

const (
	orgNameField string = "orgName"
	teamNameField string = "teamName"
)

var repoManager repomanager.RepoManager

func cloneRepo(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	var gitRepoSchema git.GitRepoSchema
	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	gitRepoSchema = git.UnmarshallGitRepoSchemaString(string(buf.Bytes()))
	if repoManager.GetOrgName() != orgName {
		http.Error(w, "Org names don't match", http.StatusBadRequest)
		return
	}
	if repoManager.ContainsGitRepoSchema(gitRepoSchema) {
		if repoManager.Update(gitRepoSchema) {
			w.WriteHeader(http.StatusOK)
			return
		} else {
			http.Error(w, "Git repo schema could not be read", http.StatusBadRequest)
			return
		}
	} else {
		if repoManager.Clone(gitRepoSchema) {
			w.WriteHeader(http.StatusOK)
			return
		} else {
			http.Error(w, "Internal server error occurred", http.StatusInternalServerError)
			return
		}
	}
}

func deleteRepo(w http.ResponseWriter, r *http.Request) {
	var gitRepoSchema git.GitRepoSchema
	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	gitRepoSchema = git.UnmarshallGitRepoSchemaString(string(buf.Bytes()))
	if repoManager.Delete(gitRepoSchema) {
		w.WriteHeader(http.StatusOK)
		return
	} else {
		http.Error(w, "Internal server error occurred", http.StatusInternalServerError)
		return
	}
}

func syncRepo(w http.ResponseWriter, r *http.Request) {
	var gitRepoSchema git.GitRepoSchema
	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	gitRepoSchema = git.UnmarshallGitRepoSchemaString(string(buf.Bytes()))
	revisionHash := repoManager.Sync(gitRepoSchema)
	if revisionHash != "" {
		w.Write([]byte(revisionHash))
		return
	} else {
		http.Error(w, "Internal server error occurred", http.StatusInternalServerError)
		return
	}
}

func resetRepoToVersion(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	orgName := vars[orgNameField]
	//teamName := vars[teamNameField]
	gitCommit := vars["gitCommit"]
	var gitRepoSchemaInfo git.GitRepoSchemaInfo
	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	gitRepoSchemaInfo = git.UnmarshallGitRepoSchemaInfoString(string(buf.Bytes()))
	gitRepoSchema := git.New(gitRepoSchemaInfo.GetGitRepo(), gitRepoSchemaInfo.GetPathToRoot(), nil)
	if repoManager.GetOrgName() == orgName && repoManager.ContainsGitRepoSchema(gitRepoSchema) {
		if repoManager.ResetToVersion(gitCommit, gitRepoSchema) {
			w.WriteHeader(http.StatusOK)
			return
		} else {
			http.Error(w, "Internal server error occurred", http.StatusInternalServerError)
			return
		}
	} else {
		http.Error(w, "Org name doesn't match or repo is not contained by manager", http.StatusBadRequest)
		return
	}
}

func InitClients(dbOperator db.DbOperator, kubernetesClient kubernetesclient.KubernetesClient) {
	orgName := os.Getenv("ORG_NAME")
	if orgName == "" {
		orgName = repomanager.DefaultOrg
	}
	repoManager = repomanager.New(dbOperator, kubernetesClient, orgName)
}

func InitRepoEndpoints(r *mux.Router) {
	r.HandleFunc("/repo/clone/{orgName}", cloneRepo).Methods("POST")
	r.HandleFunc("/repo/delete", deleteRepo).Methods("POST")
	r.HandleFunc("/repo/sync", syncRepo).Methods("POST")
	r.HandleFunc("/repo/resetToVersion/{orgName}/{teamName}/{gitCommit}", resetRepoToVersion).Methods("POST")
}
